In [225]:
import pandas as pd
import numpy as np

## Reading files

In [226]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [227]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [249]:
wh = pd.concat([wh2018,wh2019])

In [250]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [230]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

In [199]:
nulos = []
nulos1 = []
columnas = []
annios = []
df = []

In [120]:
l = wh.groupby(wh.year)

In [121]:
df1 = l.get_group(2018)

In [122]:
df2 = l.get_group(2019)

In [200]:
for c in wh.columns:
    columnas.append(c)
    annios.append(2018)
    nulos.append((np.sum(df2[c].isnull()))/df2.shape[0])


In [201]:
for c in wh.columns:
    columnas.append(c)
    annios.append(2019)
    nulos.append(np.sum(df1[c].isnull())/df1.shape[0])

In [202]:
n = list(zip(columnas,nulos,annios))

In [207]:
df3 = pd.DataFrame(n)

In [209]:
df3.columns = ["columns", "nulls", "year"]

In [216]:
df3 = df3.sort_values(["columns","year"], ascending =True)

In [215]:
df3

columns     nulls  year
19                          year  0.000000  2019
9                           year  0.000000  2018
14                Social support  0.096154  2019
4                 Social support  0.006410  2018
12                         Score  0.032051  2019
2                          Score  0.089744  2018
18     Perceptions of corruption  0.044872  2019
8      Perceptions of corruption  0.038462  2018
10                  Overall rank  0.000000  2019
0                   Overall rank  0.000000  2018
15       Healthy life expectancy  0.076923  2019
5        Healthy life expectancy  0.089744  2018
17                    Generosity  0.108974  2019
7                     Generosity  0.025641  2018
13                GDP per capita  0.057692  2019
3                 GDP per capita  0.019231  2018
16  Freedom to make life choices  0.076923  2019
6   Freedom to make life choices  0.070513  2018
11             Country or region  0.000000  2019
1              Country or region  0.000000  2018

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

In [219]:
wh.describe()

Overall rank       Score  GDP per capita  Social support  \
count    312.000000  293.000000      300.000000      296.000000   
mean      78.500000    5.392816        0.896833        1.207145   
std       45.104737    1.108112        0.399329        0.302437   
min        1.000000    2.853000        0.000000        0.000000   
25%       39.750000    4.519000        0.588500        1.047750   
50%       78.500000    5.386000        0.954000        1.265500   
75%      117.250000    6.182000        1.221500        1.455500   
max      156.000000    7.769000        2.096000        1.624000   

       Healthy life expectancy  Freedom to make life choices  Generosity  \
count               286.000000                    289.000000  291.000000   
mean                  0.659832                      0.425872    0.183309   
std                   0.256373                      0.155103    0.096846   
min                   0.000000                      0.000000    0.000000   
25%                   0.485250                      0.324000    0.110000   
50%                   0.695500                      0.450000    0.175000   
75%                   0.859750                      0.543000    0.244000   
max                   1.141000                      0.724000    0.598000   

       Perceptions of corruption         year  
count                 299.000000   312.000000  
mean                    0.110950  2018.500000  
std                     0.095574     0.500803  
min                     0.000000  2018.000000  
25%                     0.050000  2018.000000  
50%                     0.082000  2018.500000  
75%                     0.139500  2019.000000  
max                     0.457000  2019.000000

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto

### resolucion ejercicio tres

In [252]:
wh["Score"].mean()

5.392815699658706

In [253]:
for r in wh.columns:
    if r != "Country or region": 
        wh[r] = wh[r].fillna(wh[r].mean())
    count += 1

In [254]:
wh

Overall rank         Country or region     Score  GDP per capita  \
0               1                   Finland  7.632000        1.305000   
1               2                    Norway  7.594000        1.456000   
2               3                   Denmark  7.555000        1.351000   
3               4                   Iceland  7.495000        1.343000   
4               5               Switzerland  7.487000        1.420000   
5               6               Netherlands  5.392816        1.361000   
6               7                    Canada  7.328000        1.330000   
7               8               New Zealand  5.392816        1.268000   
8               9                    Sweden  7.314000        1.355000   
9              10                 Australia  7.272000        1.340000   
10             11            United Kingdom  7.190000        1.244000   
11             12                   Austria  7.139000        1.341000   
12             13                Costa Rica  7.072000        1.010000   
13             14                   Ireland  6.977000        1.448000   
14             15                   Germany  6.965000        1.340000   
15             16                   Belgium  6.927000        1.324000   
16             17                Luxembourg  6.910000        1.576000   
17             18             United States  6.886000        1.398000   
18             19                    Israel  6.814000        1.301000   
19             20      United Arab Emirates  6.774000        2.096000   
20             21            Czech Republic  6.711000        1.233000   
21             22                     Malta  6.627000        1.270000   
22             23                    France  6.489000        1.293000   
23             24                    Mexico  6.488000        1.038000   
24             25                     Chile  6.476000        1.131000   
25             26                    Taiwan  6.441000        1.365000   
26             27                    Panama  6.430000        1.112000   
27             28                    Brazil  6.419000        0.986000   
28             29                 Argentina  6.388000        1.073000   
29             30                 Guatemala  6.382000        0.781000   
30             31                   Uruguay  6.379000        0.896833   
31             32                     Qatar  6.374000        1.649000   
32             33              Saudi Arabia  6.371000        1.379000   
33             34                 Singapore  6.343000        1.529000   
34             35                  Malaysia  6.322000        1.161000   
35             36                     Spain  6.310000        1.251000   
36             37                  Colombia  6.260000        0.960000   
37             38         Trinidad & Tobago  6.192000        1.223000   
38             39                  Slovakia  6.173000        1.210000   
39             40               El Salvador  6.167000        0.806000   
40             41                 Nicaragua  6.141000        0.668000   
41             42                    Poland  6.123000        1.176000   
42             43                   Bahrain  6.105000        1.338000   
43             44                Uzbekistan  6.096000        0.719000   
44             45                    Kuwait  6.083000        1.474000   
45             46                  Thailand  6.072000        1.016000   
46             47                     Italy  6.000000        1.264000   
47             48                   Ecuador  5.973000        0.889000   
48             49                    Belize  5.956000        0.896833   
49             50                 Lithuania  5.952000        1.197000   
50             51                  Slovenia  5.948000        1.219000   
51             52                   Romania  5.945000        1.116000   
52             53                    Latvia  5.392816        1.148000   
53             54                     Japan  5.915000        1.

4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [62]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro

In [257]:
wh = wh.sort_values(["Country or region", "year"], ascending=True)

In [281]:
wh["year"].drop_duplicates().values

array([2018, 2019], dtype=int64)

In [282]:
df = []

In [284]:
for x in wh["year"].drop_duplicates().values:
    df.append(wh.loc[wh["year"] == x, :])